STUDIO/org.openl.rules.webstudio/src/org/openl/rules/ui/ProjectModel.java

```diff
@@ -1258,13 +1258,19 @@ public synchronized File getSourceByName(String fileName) {
         return null;
     }
 
-    public synchronized String getHistoryStoragePath() {
+    public String getHistoryStoragePath() {
         if (historyStoragePath == null) {
-            File location = WebStudioUtils.getUserWorkspace(WebStudioUtils.getSession())
-                .getLocalWorkspace()
-                .getLocation();
-            return Paths.get(location.getPath(), getProject().getName(), ".history", getModuleInfo().getName())
-                .toString();
+            synchronized (this) {
+                if (historyStoragePath == null) {
+                    File location = WebStudioUtils.getUserWorkspace(WebStudioUtils.getSession())
+                        .getLocalWorkspace()
+                        .getLocation();
+                    String historyFileName = getModuleInfo().getRulesRootPath().getPath();
+                    historyStoragePath = Paths
+                        .get(location.getPath(), getProject().getName(), ".history", historyFileName)
+                        .toString();
+                }
+            }
         }
         return historyStoragePath;

```

Comment:
Is it really needed lazy initialization? IMHO it can be done in the constructor.

Reply:
done